genericjpa/common-apis/src/main/java/org/hibernate/search/genericjpa/metadata/impl/RehashedTypeMetadata.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.genericjpa.metadata.impl;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+import org.hibernate.annotations.common.reflection.XProperty;
+import org.hibernate.search.backend.spi.SingularTermDeletionQuery;
+import org.hibernate.search.bridge.FieldBridge;
+import org.hibernate.search.engine.metadata.impl.TypeMetadata;
+
+/**
+ * This class represents a "rehashed" version of the
+ * TypeMetadata the engine gives us. These are mostly needed
+ * for the async backend to know where to find the id fields
+ * (on the class and in the index so the index can be held up-to-date
+ * properly) and the types of the fields
+ *
+ * @author Martin Braun
+ */
+public final class RehashedTypeMetadata {
```

Comment:
Is this shared between threads? If so, it should be immutable preferably. Btw. I found the "rehashed" name confusing, maybe "Enriched..." or "Extended..."? It's not just representing `TypeMetadata` in a different form but adds more metadata to it.